design-proposals/mass_machine_type_transition.md

```diff
@@ -0,0 +1,230 @@
+
+## Repos
+[kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+Create a new virtctl command that will allow the user to automatically update the machine type of any VMs with a machine type that is no longer supported. This command invokes a k8s `Job` that iterates through all VMs within a specified namespace (or all namespaces if none is specified), determines if the machine type is no longer supported, and updates it to the latest supported version. Both the minimum supported machine type and the target machine type are constants that can be updated as future versions are released and more machine types are deprecated upon these releases. 
+
+If the VM is running, the update will not take effect until it is restarted, so the label `restart-vm-required` will be applied to the VM until the VMI is removed (indicating the VM has been stopped), upon which the label will be cleared. The job will only terminate once all `restart-vm-required` labels have been cleared. By default, the user must handle stopping/restarting the labelled VMs manually. This is to allow the user to safely update the machine types of their VMs at a time they choose without worrying about workload interruptions. However, the user also has the option to allow the job to automatically restart every running VM that is being updated and apply the changes immediately.
+
+The options to select a specific namespace or restart the VMs immediately are configurable with the `NAMESPACE` and `RESTART_NOW` environment variables respectively.
+
+Given a VM with the following spec:
+````
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+...
+spec:
+  running: false
+  template:
+    spec:
+      domain:
+        machine:
+          type: pc-q35-rhel8.2.0
+````
+When the MMTT is invoked, the VM spec will be updated to: 
+````
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+...
+spec:
+  running: false
+  template:
+    spec:
+      domain:
+        machine:
+          type: pc-q35-rhel9.2.0
+````
+where `pc-q35-rhel9.2.0` is the latest machine type version.
+
+Along with machine types being in the format `pc-q35-rhelx.x.x`, a VM may have a machine type of `q35`, an alias for the current latest machine type version *at the time the VM is started*.
+
+For example, this VM:
+````
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+...
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        machine:
+          type: q35
+````
+may have been started when a RHEL-8 machine type was the latest machine type version and not been stopped or restarted since. Its VMI would thus look like this:
+````
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachineInstance
+...
+spec:
+  domain:
+    machine:
+      type: q35
+status:
+  machine:
+    type: pc-q35-rhel8.2.0
+````
+and need to have its machine type updated.
+
+To handle this, any running VMs with the machine type `q35` will be labelled with `restart-vm-required`:
```

Comment:
```suggestion
To handle this, any running VMs with an outdated machine type (reported in VMI's status.machine.type) will be labelled with `restart-vm-required`:
```

We should do that only for VMs with an outdated machine type.
Am I missing anything?